docs/_docs/reference/experimental/typeclasses.md

```diff
@@ -0,0 +1,672 @@
+
+---
+layout: doc-page
+title: "Type Classes"
+nightlyOf: https://docs.scala-lang.org/scala3/reference/experimental/typeclasses.html
+---
+
+# Some Proposed Changes for Better Support of Type Classes
+
+Martin Odersky, 8.1.2024
+
+A type class in Scala is a pattern where we define
+
+ - a trait with one type parameter (the _type class_)
+ - given instances at specific instantiations of that trait,
+ - using clauses or context bounds abstracting over that trait.
+
+Type classes as a pattern work overall OK, but if we compare them to native implementations in Haskell, or protocols in Swift, or traits in Rust, then there are some idiosyncracies and rough corners which in the end make them
+a bit cumbersome and limiting for standard generic programming patterns. Much has improved since Scala 2's implicits, but there is still some gap to bridge to get to parity with these languages.
+
+This note shows that with some fairly small and reasonable tweaks to Scala's syntax and typing rules we can obtain a much better scheme for working with type classes, or do generic programming in general.
+
+The bulk of the suggested improvements has been implemented and is available
+under source version `future` if the additional experimental language import `modularity` is present. For instance, using the following command:
+
+```
+  scala compile -source:future -language:experimental.modularity
+```
+
+## Generalizing Context Bounds
+
+ The only place in Scala's syntax where the type class pattern is relevant is
+ in context bounds. A context bound such as
+
+```scala
+   def min[A: Ordering](x: List[A]): A
+```
+requires that `Ordering` is a trait or class with a single type parameter (which makes it a type class) and expands to a `using` clause that instantiates that parameter. Here is the expansion of `min`:
+```scala
+   def min[A](x: List[A])(using Ordering[A]): A
+```
+
+**Proposal** Allow type classes to define an abstract type member named `Self` instead of a type parameter.
+
+**Example**
+
+```scala
+  trait Ord:
+    type Self
+
+  trait SemiGroup:
+    type Self
+    extension (x: Self) def combine(y: Self): Self
+
+  trait Monoid extends SemiGroup:
+    def unit: Self
+  object Monoid
+    def unit(using m: Monoid): m.Self = m.unit
+
+  trait Functor:
+    type Self[A]
```

Comment:
to me it's not exactly clear whether the `A` in `Self[A]` is introducing a name, or consuming a name coming from somewhere else and whether it influences e.g. extension below. maybe it's possible to clarify the exact meaning here.